### PR TITLE
Use Reference type for provider references

### PR DESCRIPTION
**Pull request by hasheddan**
### Description of your changes

We are currently using corev1.ObjectReference for provider resources.
This includes more information than needed and encourages
using helper methods that may depend on other fields in the type that we
do not intend to be utilized. This updates provider references
fields to use the Reference type, which only has a name field.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

Partial solution for #49 

Note that this will also require an update to `crossplane-tools` as mentioned in the issue.

An example of why this change is necessary is that we fail to get a provider when we use a helper function like `meta.NamespacedNameOf()` and the user has provided an extraneous `namespace` in the `providerRef`.

Slack thread: https://crossplane.slack.com/archives/CEG3T90A1/p1593112864471900

### Checklist

I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Ensured this PR contains a neat, self documenting set of commits.
- [ ] Updated any relevant [documentation], [examples], or [release notes].
- [ ] Updated the RBAC permissions in [`clusterrole.yaml`] to include any new types.

[documentation]: https://github.com/crossplane/crossplane/tree/master/docs
[examples]: https://github.com/crossplane/crossplane/tree/master/cluster/examples
[release notes]: https://github.com/crossplane/crossplane/tree/master/PendingReleaseNotes.md
[`clusterrole.yaml`]: https://github.com/crossplane/crossplane/blob/master/cluster/charts/crossplane/templates/clusterrole.yaml

### CODE REVIEW

**Comment by hasheddan** [comment]
I need to update interfaces here also

**Comment by hasheddan** [comment]
Done :heavy_check_mark:

File: apis/core/v1alpha1/resource.go [hunk 1]
```diff
@@ -197,7 +197,7 @@ type ResourceSpec struct {
 
 	// ProviderReference specifies the provider that will be used to create,
 	// observe, update, and delete this managed resource.
-	ProviderReference *corev1.ObjectReference `json:"providerRef"`
+	ProviderReference *Reference `json:"providerRef"`
```
Comment (muvaf) [inline apis/core/v1alpha1/resource.go:200]: Shouldn't this field be required actually?

```suggestion
	ProviderReference Reference `json:"providerRef"`
```

If it wasn't a pointer type, then we'd not risk having [nil pointer dereference](https://github.com/crossplane/provider-aws/pull/261) when we use it, like here in your other PR https://github.com/crossplane/provider-aws/blob/87a255e5ee6d2b2d58711f20e5d8e199101c39e2/pkg/controller/cache/cachesubnetgroup/controller.go#L86

Comment (hasheddan) [reply apis/core/v1alpha1/resource.go:200]: Great point! I'll update here :)

**Comment by hasheddan** [comment]
@muvaf updated :+1:

**Pull request merged**